File: examples/matlab/custom.m
clear;
close all;
addpath('../../build/bindings/matlab/fertilized');

% load('testdata.mat');
% Y = uint32(randi([0 1],200,1));
X = [];
Y = [];

for x = -6: .1 : 6;
    y = cos(x) * tan(x + 1);
    X = [X; [x, y]];
    Y = [Y; tan(x + 1)<0];
end

X = single(X);
Y = uint32(Y);

figure;
scatter(X(:, 1), X(:, 2), 20, Y);

s = Soil();

plotX = [-6, 6, 100];
plotY = [-6, 6, 100];

% Lets build a customized forest:
depth = 12;
n_trees = 200;
% These variables will contain the classifiers and leaf managers for each
% tree.
cls = {};
lm = {};

for i = 1:n_trees
    % StandardFeatureSelectionProvider
    stdFeatureSelect = s.StandardFeatureSelectionProvider(1, 2, 2, 2, 1+i);

    % LinearSurfaceCalculator
    linSurface = s.LinearSurfaceCalculator(400, 2, 1+i);

    % RandomizedClassificationThresholdOptimizer
    shannon = s.ShannonEntropy();
    entropyGain = s.EntropyGain(shannon);
    rcto = s.RandomizedClassificationThresholdOptimizer(1, 2, entropyGain, 0, 1, 1+i);

    % ThresholdDecider
    tClassifier = s.ThresholdDecider(stdFeatureSelect, linSurface, rcto);

    % ClassificationLeafManager
    leafMgr = s.ClassificationLeafManager(2);

    cls{i} = tClassifier;
    lm{i} = leafMgr;
end


nb = s.NoBagging();
training = s.ClassicTraining(nb);

f = s.Forest(depth, 1, 2, n_trees, cls, lm, training);

% fit forest
f.fit(X, Y);

% paint prediction plot
figure;
point_prob_plot(f, X, Y, plotX, plotY);

File: examples/python/custom.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 07 14:55:11 2014

@author: lassnech
"""
import os
import sys
sys.path.insert(0, os.path.join('..', '..', 'build', 'bindings', 'python'))
if len(sys.argv) > 1:
  print('Quiet mode. Plot display disabled.')
  INTERACTIVE = False
else:
  INTERACTIVE = True

from fertilized import *
import numpy as np
from plottools import make_spiral, point_prob_plot, accuracy_score

# Only for plotting, evaluation.
if INTERACTIVE:
    import matplotlib.pyplot as plt

np.random.seed(8)

# Generate the spiral dataset for further use.
X, Y = make_spiral()
plotx = [-6, 6, 100]
ploty = [-6, 6, 100]

if INTERACTIVE:
    plt.scatter(X[:, 0], X[:, 1], c=Y)
    plt.show()

soil = Soil('f', 'f', 'uint', Result_Types.probabilities)

###############################################################################
# Lets build a customized forest:
depth = 6
n_trees = 200
# These variables will contain the classifiers and leaf managers for each tree.
cls = []
lm = []
for i in range(n_trees):
  # Classifier:
  cls.append(soil.ThresholdDecider(
               soil.StandardFeatureSelectionProvider(1, 2, 2, 2, random_seed=1+i),
               soil.LinearSurfaceCalculator(400, random_seed=1+i),
               soil.RandomizedClassificationThresholdOptimizer(
                 True,
                 2,
                 soil.EntropyGain(soil.ShannonEntropy()), random_seed=1+i)))
  # Leaf manager:
  lm.append(soil.ClassificationLeafManager(2))

forest = soil.Forest(depth,
                     1,
                     2,
                     n_trees,
                     cls,
                     lm,
                     soil.ClassicTraining(soil.NoBagging()))
forest.fit(X, Y)
assert accuracy_score(Y, np.argmax(forest.predict(X), axis=1)) > 0.95
if INTERACTIVE:
  plt.figure()
  point_prob_plot(forest, X, Y, plotx, ploty)
  plt.savefig('custom.png')
  plt.show()
